Update multi-project docs for v1.5.0-rc1 (!)

jtcohen6 opened the pull request:
resolves https://github.com/dbt-labs/docs.getdbt.com/issues/3099 (along with https://github.com/dbt-labs/docs.getdbt.com/pull/3101)

## What are you changing in this pull request and why?

Last round of updates to docs for multi-project constructs, a.k.a. ~model publishing~ model governance

## TODOs
- [ ] Updates to model contracts, based on what gets merged by EOD tomorrow

## Checklist
<!--
Uncomment if you're publishing docs for a prerelease version of dbt (delete if not applicable):
- [x] Add versioning components, as described in [Versioning Docs](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-entire-pages)
- [ ] Add a note to the prerelease version [Migration Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/guides/migration/versions)
-->
- [x] Review the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) and [About versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) so my content adheres to these guidelines.
- [x] Add a checklist item for anything that needs to happen before this PR is merged, such as "needs technical review" or "change base branch."

Adding new pages (delete if not applicable):
- [ ] Add page to `website/sidebars.js`
- [ ] Provide a unique filename for the new page

Removing or renaming existing pages (delete if not applicable):
- [x] Remove page from `website/sidebars.js`
- [x] Add an entry `_redirects`
- [ ] [Ran link testing](https://github.com/dbt-labs/docs.getdbt.com#running-the-cypress-tests-locally) to update the links that point to the deleted page

jtcohen6 committed:
Add access to sidebars

jtcohen6 committed:
Rename publish -> governance, add redirects

jtcohen6 committed:
Fix broken links

jtcohen6 committed:
s/governance/govern

jtcohen6 committed:
Update contract example, error messages

matthewshaver commented on website/docs/docs/collaborate/govern/about-model-governance.md:
```diff
@@ -0,0 +1,11 @@
+
+:::info
+
+This section contains features new in dbt Core v1.5, currently in beta. To see the documentation related to **model groups, access, contracts, and versions**, make sure to [select v1.5 or higher](/docs/collaborate/govern/about-model-governance?version=1.5).
```
```suggestion
This section contains features new in dbt Core v1.5. To see the documentation related to **model groups, access, contracts, and versions**, make sure to [select v1.5 or higher](/docs/collaborate/govern/about-model-governance?version=1.5).
```

MichelleArk committed:
model versions (#3101)

Addresses: https://github.com/dbt-labs/docs.getdbt.com/issues/3099

## What are you changing in this pull request and why?
<!---
Describe your changes and why you're making them. If linked to an open
issue or a pull request on dbt Core, then link to them here! 

To learn more about the writing conventions used in the dbt Labs docs,
see the [Content style
guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md).
-->

## Checklist
<!--
Uncomment if you're publishing docs for a prerelease version of dbt
(delete if not applicable):
- [ ] Add versioning components, as described in [Versioning
Docs](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/versioningdocs.md)
- [ ] Add a note to the prerelease version [Migration
Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/guides/migration/versions)
-->
- [ ] Review the [Content style
guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md)
and [About
versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version)
so my content adheres to these guidelines.
- [ ] Add a checklist item for anything that needs to happen before this
PR is merged, such as "needs technical review" or "change base branch."

Adding new pages (delete if not applicable):
- [ ] Add page to `website/sidebars.js`
- [ ] Provide a unique filename for the new page

Removing or renaming existing pages (delete if not applicable):
- [ ] Remove page from `website/sidebars.js`
- [ ] Add an entry `_redirects`
- [ ] [Ran link
testing](https://github.com/dbt-labs/docs.getdbt.com#running-the-cypress-tests-locally)
to update the links that point to the deleted page

---------

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>
Co-authored-by: Leona B. Campbell <3880403+runleonarun@users.noreply.github.com>
Co-authored-by: mirnawong1 <89008547+mirnawong1@users.noreply.github.com>
Co-authored-by: Jeremy Cohen <jeremy@dbtlabs.com>

jtcohen6 committed:
Lots of updates

jtcohen6 committed:
Merge branch 'current' into jerco/multi-project-last-round

jtcohen6 committed:
More updates from self-review

matthewshaver committed:
Update _redirects

matthewshaver commented on website/docs/docs/build/custom-aliases.md:
```diff
@@ -114,3 +147,21 @@ Compilation Error
+- [Model versions](govern/model-versions)
+- [`versions`](resource-properties/versions#alias)
+
+By default, dbt will create versioned models with the alias `<model_name>_v<v>`, where `<v>` is that version's unique identifier. You can customize this behavior just like you can for non-versioned models, by configuring a custom `alias` or reimplementing the `generate_alias_name` macro.
```
```suggestion
By default, dbt will create versioned models with the alias `<model_name>_v<v>`, where `<v>` is that version's unique identifier. You can customize this behavior just like for non-versioned models by configuring a custom `alias` or re-implementing the `generate_alias_name` macro.
```

matthewshaver committed:
Update website/docs/docs/build/custom-aliases.md

matthewshaver commented on website/docs/docs/collaborate/govern/about-model-governance.md:
```diff
@@ -0,0 +1,17 @@
+
+:::
+
+[**Model access**](model-access): Some models are mature, reusable data productions. Others are your team's implementation details on the way there. Mark models as "public" or "private," to make the distinction clear, and to control who else can `ref` them.
```
```suggestion
[**Model access**](model-access): Some models are mature, reusable data productions. Others are your team's implementation details on the way there. Mark models as "public" or "private," to make the distinction clear and to control who else can `ref` them.
```

matthewshaver committed:
Update website/docs/docs/collaborate/govern/about-model-governance.md

matthewshaver commented on website/docs/docs/collaborate/govern/about-model-governance.md:
```diff
@@ -0,0 +1,17 @@
+
+[**Model access**](model-access): Some models are mature, reusable data productions. Others are your team's implementation details on the way there. Mark models as "public" or "private," to make the distinction clear and to control who else can `ref` them.
+
+[**Model contracts**](model-contracts): Guarantee the shape of a model, while it is building, to avoid surprises—or breaking changes for downstream queriers. Explicitly define column names, data types, and constraints (as supported by your data platform).
```
```suggestion
[**Model contracts**](model-contracts): Guarantee the shape of a model while it is building to avoid surprises or breaking changes for downstream queries. Explicitly define column names, data types, and constraints (as supported by your data platform).
```

matthewshaver committed:
Update website/docs/docs/collaborate/govern/about-model-governance.md

matthewshaver commented on website/docs/docs/collaborate/govern/about-model-governance.md:
```diff
@@ -0,0 +1,17 @@
+
+[**Model contracts**](model-contracts): Guarantee the shape of a model while it is building to avoid surprises or breaking changes for downstream queries. Explicitly define column names, data types, and constraints (as supported by your data platform).
+
+[**Model versions**](model-versions): When a breaking change is unavoidable, provide a smoother upgrade pathway by creating a new version of the model. These model versions share a common reference name, and can reuse properties & configuration.
```
```suggestion
[**Model versions**](model-versions): When a breaking change is unavoidable, provide a smoother upgrade pathway by creating a new version of the model. These model versions share a common reference name and can reuse properties & configurations.
```

matthewshaver committed:
Update website/docs/docs/collaborate/govern/about-model-governance.md

matthewshaver commented on website/docs/docs/collaborate/govern/model-versions.md:
```diff
@@ -0,0 +1,116 @@
+
+[Version control](git-version-control) allows your team to collaborate simultaneously on a single code repository, manage conflicts between changes, and review changes before deploying into production. In that sense, version control is an essential tool for versioning the deployment of an entire dbt project—always the latest state of the `main` branch, with the ability to "rollback" changes by re ...
+
+Model versions are different. Multiple versions of a model will live in the same code repository at the same time, and be deployed into the same data environment simultaneously. This is similar to how web APIs are versioned—multiple versions are live at the same time; older versions are often eventually sunsetted.
```
```suggestion
Model versions are different. Multiple versions of a model will live in the same code repository at the same time and be deployed into the same data environment simultaneously. This is similar to how web APIs are versioned—multiple versions are live simultaneously; older versions are often eventually sunsetted.
```

matthewshaver committed:
Update website/docs/docs/collaborate/govern/model-versions.md

matthewshaver commented on website/docs/docs/collaborate/govern/model-versions.md:
```diff
@@ -0,0 +1,116 @@
+
+</File>
+
+If you wanted to make a breaking change to the model - for example, removing a column - you'd create a new model file (SQL or python file) encompassing those breaking changes. The default convention is naming the new file with a `_v<version>` suffix. The new version can then be configured in relation to the original model:
```
```suggestion
If you wanted to make a breaking change to the model - for example, removing a column - you'd create a new model file (SQL or Python) encompassing those breaking changes. The default convention is naming the new file with a `_v<version>` suffix. The new version can then be configured in relation to the original model:
```

matthewshaver commented on website/docs/docs/collaborate/govern/model-versions.md:
````diff
@@ -0,0 +1,116 @@
+```
+
+:::info
+Projects which have historically implemented [custom aliases](/docs/build/custom-aliases) by reimplemented the `generate_alias_name` macro will need to update their custom implementations to account for model versions. 
````
```suggestion
Projects which have historically implemented [custom aliases](/docs/build/custom-aliases) by reimplementing the `generate_alias_name` macro will need to update their custom implementations to account for model versions. 
```

matthewshaver commented on website/docs/docs/collaborate/govern/model-versions.md:
```diff
@@ -0,0 +1,116 @@
+:::info
+Projects which have historically implemented [custom aliases](/docs/build/custom-aliases) by reimplemented the `generate_alias_name` macro will need to update their custom implementations to account for model versions. 
+
+Otherwise they'll see something like this as soon as they start using versions:
```
```suggestion
Otherwise, they'll see something like this as soon as they start using versions:
```

matthewshaver commented on website/docs/reference/dbt-jinja-functions/ref.md:
```diff
@@ -36,6 +36,40 @@ The `{{ ref }}` function returns a `Relation` object that has the same `table`,
+
+This functionality is useful when referencing versioned models that make breaking changes by creating new versions, but guaruntee no breaking changes to existing versions of the model.
+
+If the `version` argument is not supplied to a `ref` to a versioned model, the latest version. This has the benefit of automatically incorporating the latest changes of a referenced model, but is at risk to incorporating breaking changes.
```
```suggestion
If the `version` argument is not supplied to a `ref` of a versioned model, the latest version is. This has the benefit of automatically incorporating the latest changes of a referenced model, but there is a risk of incorporating breaking changes.
```

matthewshaver commented on website/docs/reference/node-selection/methods.md:
```diff
@@ -275,6 +275,28 @@ The `group` method is used to select models defined within a group.
+
+<VersionBlock firstVersion="1.5">
+
+The `version` method selects [versioned models](model-versions), based on their [version identifier](resource-properties/versions) and [latest version](resource-properties/latest_version).
```
```suggestion
The `version` method selects [versioned models](model-versions) based on their [version identifier](resource-properties/versions) and [latest version](resource-properties/latest_version).
```

matthewshaver commented on website/docs/reference/resource-properties/constraints.md:
```diff
@@ -3,14 +3,62 @@ resource_types: [models]
+This functionality is new in v1.5.
 :::
 
+Constraints are a feature of many data platforms. When specified, the platform will perform additional validation on data as it is being populated in a new table, or inserted into a preexisting table. If the validation fails, the table creation or update fails, the operation is rolled back, and you will see a clear error message.
```
```suggestion
Constraints are a feature of many data platforms. When specified, the platform will perform additional validation on data as it is being populated in a new table or inserted into a preexisting table. If the validation fails, the table creation or update fails, the operation is rolled back, and you will see a clear error message.
```

matthewshaver commented on website/docs/reference/resource-properties/constraints.md:
```diff
@@ -3,14 +3,62 @@ resource_types: [models]
+
+Constraints require the declaration and enforcement of a model [contract](resource-configs/contract)
+
+**Constraints are never applied on models materialized as `view`** (or `ephemeral`). Only `table` and `incremental` models support the application and enforcement of constraints.
```
```suggestion
**Constraints are never applied on `ephemeral` models or those materialized as `view`**. Only `table` and `incremental` models support applying and enforcing constraints.
```

matthewshaver commented on website/docs/reference/resource-properties/include-exclude.md:
```diff
@@ -0,0 +1,105 @@
+</File>
+
+## Definition
+The specification of which columns defined in a model's top-level `columns` property to include or exclude in a versioned implementation of that model.
```
```suggestion
The specification of which columns are defined in a model's top-level `columns` property to include or exclude in a versioned implementation of that model.
```

matthewshaver commented on website/docs/reference/resource-properties/include-exclude.md:
```diff
@@ -0,0 +1,105 @@
+
+Each other version has declared a modification from the top-level property:
+- `v3` will include all columns, but it reimplements the `customer_country` column with a different `description`
+- `v2` will include all columns *except* `customer_country`
```
```suggestion
- `v2` will include all columns _except_ `customer_country`
```

matthewshaver commented on website/docs/reference/resource-properties/latest-version.md:
```diff
@@ -0,0 +1,68 @@
+1. Resolving `ref()` calls to this model that are "unpinned" (a version is not explicitly specified)
+2. Selecting model versions using the [`version:` selection method](methods#the-version-method), based on whether a given model version is `latest`, `prerelease`, or `old`
+
+This value can be either a string or numeric (integer or float) value, and must be one of the [version identifiers](resource-properties/versions#v) specified in this model's list of `versions`.
```
```suggestion
This value can be a string or a numeric (integer or float) value. It must be one of the [version identifiers](resource-properties/versions#v) specified in this model's list of `versions`.
```

matthewshaver commented on website/docs/reference/resource-properties/latest-version.md:
```diff
@@ -0,0 +1,68 @@
+
+For a non-versioned model (no `versions` list), `latest_version` has no value.
+
+If `latest_version` is not specified for a versioned model, `latest_version` defaults to the largest 
```
```suggestion
If `latest_version` is not specified for a versioned model, `latest_version` defaults to the largest.
```

matthewshaver commented on website/docs/reference/resource-properties/versions.md:
```diff
@@ -0,0 +1,58 @@
+
+The name of the model file (excluding the file extension, e.g. `.sql` or `.py`) where the model version is defined.
+
+If `defined_in` is not specified, dbt searches for the definition of a versioned model in a model file named `<model_name>_v<v>`. Model file names must be globally unique, even when they are defining versioned implementations of a model with a different name.
```
```suggestion
If `defined_in` is not specified, dbt searches for the definition of a versioned model in a model file named `<model_name>_v<v>`. Model file names must be globally unique, even when defining versioned implementations of a model with a different name.
```

matthewshaver committed:
Apply suggestions from code review

matthewshaver committed:
Merge branch 'current' into jerco/multi-project-last-round
